### PR TITLE
Bug 2082428: Add MicroSecond processing

### PR DESCRIPTION
**Pull request by candita**
Corresponds to openshift/api#1197

Valid time units for duration are "ns", "us" (or "µs"), "ms", "s", "m", "h"."  
Added ability to process microseconds so that fractional milliseconds (1.5ms = 1500us) could be represented, but did not add ability to process nanoseconds because it is uncommon to need fractional microseconds.

Also bumped api and added a few unit tests.

### CODE REVIEW

**Comment by candita** [comment]
>   host command failed: error running /usr/bin/kubectl --server=https://api.ci-op-ygg84x42-2945d.origin-ci-int-aws.dev.rhcloud.com:6443 --kubeconfig=/tmp/kubeconfig-2432296801 --namespace=e2e-test-unprivileged-router-lcw7t exec execpod -- /bin/sh -x -c 
>     		set -e
>     		pass=5
>     		STOP=$(($(date '+%s') + 180))
>     		while [ $(date '+%s') -lt $STOP ]; do
>     			rc=0
>     			code=$( curl -k -s -m 5 -o /dev/null -w '%{http_code}\n' --header 'Host: 10.131.0.179' "http://10.131.0.179:1936/healthz" ) || rc=$?
>     			if [[ "${rc:-0}" -eq 0 ]]; then
>     				echo $code
>     				if [[ $code -eq 200 ]]; then
>     					pass=$(( pass - 1 ))
>     					if [[ $pass -le 0 ]]; then
>     						exit 0
>     					fi
>     					sleep 1
>     					continue
>     				fi
>     				if [[ $code -ne 503 ]]; then
>     					exit 1
>     				fi
>     				pass=5
>     			else
>     				echo "error ${rc}" 1>&2
>     			fi
>     			sleep 1
>     		done
>     		:
>     Command stdout:
>     500

/test e2e-aws

**Comment by candita** [comment]
> It] alert/etcdHighNumberOfFailedGRPCRequests should not be at or above info [Suite:openshift/conformance/parallel]
>   github.com/openshift/origin/test/extended/prometheus/alerts.go:23
> 
/retest

**Comment by candita** [comment]
> alert HighOverallControlPlaneCPU fired for 690 seconds with labels: {namespace="openshift-kube-apiserver", severity="warning"} (allowed: high CPU utilization during e2e runs is normal)

/retest

File: pkg/operator/controller/ingress/deployment.go [hunk 1]
```diff
@@ -1642,11 +1642,19 @@ func clipHAProxyTimeoutValue(val string) (string, error) {
...
+	} else if durationMicroseconds != 0 && durationMilliseconds == 0 {
+		return fmt.Sprintf("%dus", durationMicroseconds)
+	} else if durationMicroseconds > haproxyMaxTimeoutMicroseconds {
 		log.V(2).Info("Time value %s exceeds the maximum timeout length of %dms. Truncating timeout to match maximum value.", duration.String(), haproxyMaxTimeoutMilliseconds)
 		return "2147483647ms"
+	} else if durationMicroseconds%time.Millisecond.Microseconds() != 0 {
+		return fmt.Sprintf("%dus", durationMicroseconds)
 	} else if durationMilliseconds%time.Second.Milliseconds() != 0 {
 		return fmt.Sprintf("%dms", durationMilliseconds)
 	} else if durationMilliseconds%time.Minute.Milliseconds() != 0 {
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment.go:1660]: Is this simpler?
```go
const haproxyMaxTimeout = 2147483647 * time.Millisecond

func durationToHAProxyTimespec(duration time.Duration) string {
	if duration <= 0 {
		return "0s"
	} else if duration > haproxyMaxTimeout {
		log.V(2).Info("time value %v exceeds the maximum timeout length of %v; truncating to maximum value", duration, haproxyMaxTimeout)
		return "2147483647ms"
	} else if µs := duration.Microseconds(); µs%1000 != 0 {
		return fmt.Sprintf("%dus", µs)
	} else if ms := duration.Milliseconds(); ms%1000 != 0 {
		return fmt.Sprintf("%dms", ms)
	} else if ms%time.Minute.Milliseconds() != 0 {
		return fmt.Sprintf("%ds", int(math.Round(duration.Seconds())))
	} else if ms%time.Hour.Milliseconds() != 0 {
		return fmt.Sprintf("%dm", int(math.Round(duration.Minutes())))
	} else {
		return fmt.Sprintf("%dh", int(math.Round(duration.Hours())))
	}
}
```

Comment (candita) [reply pkg/operator/controller/ingress/deployment.go]: Yes, added.

**Comment by candita** [comment]
> [It] alert/etcdHighNumberOfFailedGRPCRequests should not be at or above info [Suite:openshift/conformance/parallel]
>   github.com/openshift/origin/test/extended/prometheus/alerts.go:23
> May 27 20:27:20.235: FAIL: etcdHighNumberOfFailedGRPCRequests was at or above info for at least 2m29s on platformidentification.JobType{Release:"4.11", FromRelease:"", Platform:"gcp", Architecture:"amd64", Network:"sdn", Topology:"ha"} (maxAllowed=3s): pending for 16m27s, firing for 2m29s:
> 

/test e2e-gcp-serial

**Comment by Miciah** [comment]
Thanks!
/lgtm

**Pull request merged**